tests/nc-client/main.c

```diff
@@ -0,0 +1,157 @@
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "events.h"
+#include "network.h"
+#include "sock.h"
+#include "warnp.h"
+
+struct senddata {
+	char * buffer;
+	int socket;
+	int conndone;
+	void * connect_cookie;
+	void * write_cookie;
+};
+
+/* Forward declarations. */
+static int send_input(void *);
+static int callback_wrote(void * cookie, ssize_t lenwrit);
+static int callback_connected(void * cookie, int socket);
+
+/* Send data from stdin to a socket, or close the connection. */
+static int
+send_input(void * cookie)
+{
+	struct senddata * send = (struct senddata *)cookie;
+	size_t len = 0;
+	ssize_t read;
+
+	/* Read data from stdin. */
+	if ((read = getline(&send->buffer, &len, stdin)) != -1) {
+		/* Send data to server. */
+		if ((send->write_cookie = network_write(send->socket,
+		    (uint8_t *)send->buffer, (size_t)read, (size_t)read,
+		    callback_wrote, cookie)) == NULL) {
+			warn0("network_write failure");
+			goto err0;
+		}
+	} else {
+		/* Close connection without sending anything. */
+		send->conndone = 1;
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (1);
+}
+
+/* Finished writing data; look for more from stdin. */
+static int
+callback_wrote(void * cookie, ssize_t lenwrit)
+{
+	struct senddata * send = (struct senddata *)cookie;
+
+	(void)lenwrit; /* UNUSED */
```

Comment:
Maybe you want to check `lenwrit`?  If the destination process closes the socket before all the data is sent, do you want nc_client to report an error?